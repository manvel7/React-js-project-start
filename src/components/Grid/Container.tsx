import {HTMLAttributes} from "react"
import clsx from "clsx"
import _ from "lodash"
import styled, {css, StyledProps} from "styled-components"
import {ColBreakpointName, ScreenBreakpointName} from "../../interfaces/grid"
import {mediaBreakpointUp} from "../../utils/breakpoint"
import {pxToRem} from "../../utils/unit"

export interface IProps extends HTMLAttributes<HTMLDivElement> {
  fluid?: boolean | ColBreakpointName
}

const maxWidthMixin = css<StyledProps<IProps>>`
  ${({
    fluid,
    theme: {
      grid: {breakpoints, containerMaxWidths},
      typography: {htmlFontSize}
    }
  }) => {
    if (fluid === true) {
      return ""
    }

    let shouldSkip = false

    return (Object.keys(breakpoints) as Array<ScreenBreakpointName>)
      .map((breakpoint) => {
        if (breakpoint === fluid) {
          shouldSkip = true
        }

        const maxWidth = containerMaxWidths[breakpoint]

        if (maxWidth) {
          return mediaBreakpointUp(
            breakpoint,
            breakpoints,
            shouldSkip
              ? "max-width: none;"
              : `max-width: ${pxToRem(maxWidth, htmlFontSize)};`
          )
        }

        return ""
      })
      .join("\n")
  }}
`

const Container = styled.div.attrs<IProps, IProps>(({className, fluid}) => {
  return {
    className: clsx(className, {
      container: true,
      "container-fluid": !_.isNil(fluid) && fluid === true,
      [`container-${fluid}-fluid`]: !_.isNil(fluid) && fluid !== true
    })
  }
})`
  margin-left: auto;
  margin-right: auto;
  padding-left: calc(
    ${({
        theme: {
          grid: {gutterWidth},
          typography: {htmlFontSize}
        }
      }) => pxToRem(gutterWidth, htmlFontSize)} / 2
  );
  padding-right: calc(
    ${({
        theme: {
          grid: {gutterWidth},
          typography: {htmlFontSize}
        }
      }) => pxToRem(gutterWidth, htmlFontSize)} / 2
  );
  width: 100%;

  ${maxWidthMixin}
`

export default Container
