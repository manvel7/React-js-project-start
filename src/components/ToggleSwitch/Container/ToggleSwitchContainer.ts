import {LabelHTMLAttributes} from "react"
import styled from "styled-components"

interface IProps extends LabelHTMLAttributes<HTMLLabelElement> {}

const ToggleSwitchContainer = styled.label<IProps>`
  align-items: center;
  cursor: pointer;
  display: inline-flex;
  justify-content: center;
  padding: 0.8rem;
`

export default ToggleSwitchContainer
