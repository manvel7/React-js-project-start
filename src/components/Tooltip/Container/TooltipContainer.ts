import styled from "styled-components"

const TooltipContainer = styled.div`
  background-color: ${({theme}) => theme.palette.background.white};
  border: 0.1rem solid ${({theme}) => theme.palette.border};
  border-radius: 0.3rem;
  display: flex;
  flex-direction: column;
  max-width: 20rem;
  padding: 0.8rem 1.6rem;
  z-index: 99;
`

export default TooltipContainer
