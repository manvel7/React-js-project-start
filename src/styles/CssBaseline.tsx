import {
  createGlobalStyle,
  css,
  DefaultTheme,
  GlobalStyleComponent,
  StyledProps
} from "styled-components"
import {mediaBreakpointDown} from "../utils/breakpoint"
import {getValue, toPercent} from "../utils/unit"
import {
  bodyMixin,
  smallMixin,
  h1Mixin,
  h2Mixin,
  h3Mixin,
  h4Mixin,
  h5Mixin,
  h6Mixin
} from "./typography"

interface IProps {}

const genericResetMixin = css<StyledProps<IProps>>`
  html,
  body,
  div,
  span,
  applet,
  object,
  iframe,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6,
  p,
  blockquote,
  pre,
  a,
  abbr,
  acronym,
  address,
  big,
  cite,
  code,
  del,
  dfn,
  em,
  img,
  ins,
  kbd,
  q,
  s,
  samp,
  small,
  strike,
  strong,
  sub,
  sup,
  tt,
  var,
  b,
  u,
  i,
  center,
  dl,
  dt,
  dd,
  ol,
  ul,
  li,
  fieldset,
  form,
  label,
  legend,
  table,
  caption,
  tbody,
  tfoot,
  thead,
  tr,
  th,
  td,
  article,
  aside,
  canvas,
  details,
  embed,
  figure,
  figcaption,
  footer,
  header,
  hgroup,
  menu,
  nav,
  output,
  ruby,
  section,
  summary,
  time,
  mark,
  audio,
  video {
    margin: 0;
    padding: 0;
    border: 0;
    vertical-align: baseline;
  }

  article,
  aside,
  details,
  figcaption,
  figure,
  footer,
  header,
  hgroup,
  menu,
  nav,
  section {
    display: block;
  }

  ol,
  ul {
    list-style: none;
  }

  blockquote,
  q {
    quotes: none;
  }

  blockquote:before,
  blockquote:after,
  q:before,
  q:after {
    content: "";
    content: none;
  }

  table {
    border-collapse: collapse;
    border-spacing: 0;
  }
`

const appResetMixin = css<StyledProps<IProps>>`
  html {
    /* stylelint-disable property-no-vendor-prefix */
    -ms-overflow-style: scrollbar; /* Set box-sizing so that the width is not affected by padding or border */
    -ms-text-size-adjust: 100%; /* Set default font family in all browsers. */
    -webkit-tap-highlight-color: rgba(0, 0, 0, 0); /* Makes 1rem = 10px */
    -webkit-text-size-adjust: 100%; /* Prevent adjustments of font size after orientation changes in IE on Windows Phone and in iOS. */
    /* stylelint-enable property-no-vendor-prefix */
    box-sizing: border-box;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }

  body {
    background-color: ${({theme}) => theme.palette.background.white};

    /* Hide overflow content */
    &.overflow {
      overflow: hidden;
    }

    /* Prevent focus outline on elements that can't be accessed via keyboard. */
    [tabindex="-1"]:focus {
      outline: none !important;
    }

    /* Hide overflow in IE */
    svg:not(:root) {
      overflow: hidden;
    }

    /* Hide elements has 'hidden' attribute in IE 10- */
    [hidden] {
      display: none !important;
    }

    ::-webkit-file-upload-button {
      /* stylelint-disable-next-line property-no-vendor-prefix */
      -webkit-appearance: button;
      font: inherit;
    }
  }
`

const typographyMixin = css<StyledProps<IProps>>`
  html {
    font-family: ${({theme}) =>
      theme.typography
        .fontFamily}; /* Prevent scrollbars to overlap content in IE11 and Edge */
    font-size: ${({theme}) =>
      toPercent(
        getValue(theme.typography.htmlFontSize),
        16
      )}; /* Set default tap highlight to be transparent in iOS. */
    font-style: normal;
    color: ${({theme}) => theme.palette.text.dark};
  }

  body {
    ${bodyMixin<IProps>()}
  }

  h1 {
    ${h1Mixin<IProps>()}
  }

  h2 {
    ${h2Mixin<IProps>()}
  }

  h3 {
    ${h3Mixin<IProps>()}
  }

  h4 {
    ${h4Mixin<IProps>()}
  }

  h5 {
    ${h5Mixin<IProps>()}
  }

  h6 {
    ${h6Mixin<IProps>()}
  }

  small {
    ${smallMixin<IProps>()}
  }

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.bold};

    h1,
    h2,
    h3,
    h4,
    h5,
    h6,
    caption {
      font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
    }
  }
`

const utilsMixin = css<StyledProps<IProps>>`
  .hidden {
    visibility: hidden;
  }

  .d-none {
    display: none;
  }

  .hidden-sm-down {
    ${({theme}) =>
      mediaBreakpointDown(
        "sm",
        theme.grid.breakpoints,
        `
      display: none;
    `
      )}
  }

  .text-center {
    text-align: center;
  }
`

const CssBaseline: GlobalStyleComponent<
  IProps,
  DefaultTheme
> = createGlobalStyle`
  ${genericResetMixin}

  ${appResetMixin}

  ${typographyMixin}

  ${utilsMixin}
`

export default CssBaseline
