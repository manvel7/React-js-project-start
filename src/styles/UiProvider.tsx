import React, {FunctionComponent} from "react"
import {ThemeProvider} from "styled-components"
import theme from "./theme"
import {CssBaseline} from "./index"

interface IProps {}

const UiProvider: FunctionComponent<IProps> = (props) => {
  const {children} = props

  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      {children}
    </ThemeProvider>
  )
}

export default UiProvider
