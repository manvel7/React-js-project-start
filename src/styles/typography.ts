import {
  css,
  DefaultTheme,
  FlattenInterpolation,
  StyledProps,
  ThemedStyledProps
} from "styled-components"
import {pxToRem} from "../utils/unit"

const h1Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h1.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h1.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h1.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h1.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h2Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h2.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h2.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h2.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h2.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h3Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h3.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h3.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h3.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h3.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h4Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h4.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h4.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h4.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h4.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h5Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h5.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h5.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h5.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h5.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h6Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h6.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h6.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h6.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h6.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const bodyMixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(
      theme.typography.variants.body.size,
      theme.typography.htmlFontSize
    )};
  font-weight: ${({theme}) => theme.typography.variants.body.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.body.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.body.lineHeight,
      theme.typography.htmlFontSize
    )};
`

const smallMixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(
      theme.typography.variants.caption.size,
      theme.typography.htmlFontSize
    )};
  font-weight: ${({theme}) => theme.typography.variants.caption.weight};
  letter-spacing: ${({theme}) =>
    theme.typography.variants.caption.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.caption.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`
export {
  h1Mixin,
  h2Mixin,
  h3Mixin,
  h4Mixin,
  h5Mixin,
  h6Mixin,
  bodyMixin,
  smallMixin
}
