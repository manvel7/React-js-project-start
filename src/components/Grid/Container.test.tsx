import React from "react"
import {render} from "@testing-library/react"
import {theme} from "../../styles"
import Container from "./Container"

describe("Container", () => {
  describe("default", () => {
    it("should match snapshot", () => {
      const result = render(
        <Container theme={theme}>This is a container</Container>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })
  })

  describe("fluid", () => {
    it("on all breakpoints should match snapshot", () => {
      const result = render(
        <Container theme={theme} fluid={true}>
          This is a container
        </Container>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })

    it("on xs breakpoint should match snapshot", () => {
      const result = render(
        <Container theme={theme} fluid="xs">
          This is a container
        </Container>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })

    it("on sm breakpoint should match snapshot", () => {
      const result = render(
        <Container theme={theme} fluid="sm">
          This is a container
        </Container>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })

    it("on md breakpoint should match snapshot", () => {
      const result = render(
        <Container theme={theme} fluid="md">
          This is a container
        </Container>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })

    it("on lg breakpoint should match snapshot", () => {
      const result = render(
        <Container theme={theme} fluid="lg">
          This is a container
        </Container>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })

    it("on xl breakpoint should match snapshot", () => {
      const result = render(
        <Container theme={theme} fluid="xl">
          This is a container
        </Container>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })
  })
})
