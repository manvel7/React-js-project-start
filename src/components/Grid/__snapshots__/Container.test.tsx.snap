// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container default should match snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  padding-left: calc( 3.2rem / 2 );
  padding-right: calc( 3.2rem / 2 );
  width: 100%;
}

@media (min-width:576px) {
  .c0 {
    max-width: 54.4rem;
  }
}

@media (min-width:768px) {
  .c0 {
    max-width: 72rem;
  }
}

@media (min-width:992px) {
  .c0 {
    max-width: 96rem;
  }
}

@media (min-width:1200px) {
  .c0 {
    max-width: 114.4rem;
  }
}

<div
  class="c0 container"
>
  This is a container
</div>
`;

exports[`Container fluid on all breakpoints should match snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  padding-left: calc( 3.2rem / 2 );
  padding-right: calc( 3.2rem / 2 );
  width: 100%;
}

<div
  class="c0 container container-fluid"
>
  This is a container
</div>
`;

exports[`Container fluid on lg breakpoint should match snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  padding-left: calc( 3.2rem / 2 );
  padding-right: calc( 3.2rem / 2 );
  width: 100%;
}

@media (min-width:576px) {
  .c0 {
    max-width: 54.4rem;
  }
}

@media (min-width:768px) {
  .c0 {
    max-width: 72rem;
  }
}

@media (min-width:992px) {
  .c0 {
    max-width: none;
  }
}

@media (min-width:1200px) {
  .c0 {
    max-width: none;
  }
}

<div
  class="c0 container container-lg-fluid"
>
  This is a container
</div>
`;

exports[`Container fluid on md breakpoint should match snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  padding-left: calc( 3.2rem / 2 );
  padding-right: calc( 3.2rem / 2 );
  width: 100%;
}

@media (min-width:576px) {
  .c0 {
    max-width: 54.4rem;
  }
}

@media (min-width:768px) {
  .c0 {
    max-width: none;
  }
}

@media (min-width:992px) {
  .c0 {
    max-width: none;
  }
}

@media (min-width:1200px) {
  .c0 {
    max-width: none;
  }
}

<div
  class="c0 container container-md-fluid"
>
  This is a container
</div>
`;

exports[`Container fluid on sm breakpoint should match snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  padding-left: calc( 3.2rem / 2 );
  padding-right: calc( 3.2rem / 2 );
  width: 100%;
}

@media (min-width:576px) {
  .c0 {
    max-width: none;
  }
}

@media (min-width:768px) {
  .c0 {
    max-width: none;
  }
}

@media (min-width:992px) {
  .c0 {
    max-width: none;
  }
}

@media (min-width:1200px) {
  .c0 {
    max-width: none;
  }
}

<div
  class="c0 container container-sm-fluid"
>
  This is a container
</div>
`;

exports[`Container fluid on xl breakpoint should match snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  padding-left: calc( 3.2rem / 2 );
  padding-right: calc( 3.2rem / 2 );
  width: 100%;
}

@media (min-width:576px) {
  .c0 {
    max-width: 54.4rem;
  }
}

@media (min-width:768px) {
  .c0 {
    max-width: 72rem;
  }
}

@media (min-width:992px) {
  .c0 {
    max-width: 96rem;
  }
}

@media (min-width:1200px) {
  .c0 {
    max-width: none;
  }
}

<div
  class="c0 container container-xl-fluid"
>
  This is a container
</div>
`;

exports[`Container fluid on xs breakpoint should match snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  padding-left: calc( 3.2rem / 2 );
  padding-right: calc( 3.2rem / 2 );
  width: 100%;
}

@media (min-width:576px) {
  .c0 {
    max-width: none;
  }
}

@media (min-width:768px) {
  .c0 {
    max-width: none;
  }
}

@media (min-width:992px) {
  .c0 {
    max-width: none;
  }
}

@media (min-width:1200px) {
  .c0 {
    max-width: none;
  }
}

<div
  class="c0 container container-xs-fluid"
>
  This is a container
</div>
`;
